In [2]:
%load_ext sql

In [3]:
%sql mysql+pymysql://root:12345678@localhost:3306/brazil

'Connected: root@brazil'

In [4]:
%sql SHOW TABLES

 * mysql+pymysql://root:***@localhost:3306/brazil
9 rows affected.


[('category_translation',),
 ('customers',),
 ('geolocation',),
 ('order_items',),
 ('order_payments',),
 ('order_reviews',),
 ('orders',),
 ('products',),
 ('sellers',)]

### <font color='red'>Desired Schema <font/>

<img src="https://i.imgur.com/HRhd2Y0.png" alt="Desired Schema"/>


### ٌRenaming columns to replace spaces with underscores

In [19]:
%%sql
-- using a sql query to write the renaming query
SELECT CONCAT('ALTER TABLE ', TABLE_NAME, ' RENAME COLUMN ', '`',COLUMN_NAME,'`', ' TO ', REPLACE(COLUMN_NAME, ' ', '_'), ';')
FROM INFORMATION_SCHEMA.COLUMNS
WHERE TABLE_SCHEMA = 'brazil';

 * mysql+pymysql://root:***@localhost:3306/brazil
50 rows affected.


[('ALTER TABLE category_translation RENAME COLUMN `product_category_name` TO product_category_name;',),
 ('ALTER TABLE category_translation RENAME COLUMN `product_category_name_english` TO product_category_name_english;',),
 ('ALTER TABLE customers RENAME COLUMN `customer_city` TO customer_city;',),
 ('ALTER TABLE customers RENAME COLUMN `customer_id` TO customer_id;',),
 ('ALTER TABLE customers RENAME COLUMN `customer_state` TO customer_state;',),
 ('ALTER TABLE customers RENAME COLUMN `customer_unique_id` TO customer_unique_id;',),
 ('ALTER TABLE customers RENAME COLUMN `customer_zip_code_prefix` TO customer_zip_code_prefix;',),
 ('ALTER TABLE geolocation RENAME COLUMN `geolocation city` TO geolocation_city;',),
 ('ALTER TABLE geolocation RENAME COLUMN `geolocation state` TO geolocation_state;',),
 ('ALTER TABLE geolocation RENAME COLUMN `geolocation zip code prefix` TO geolocation_zip_code_prefix;',),
 ('ALTER TABLE order_items RENAME COLUMN `freight value` TO freight_value;',),
 ('ALTER TABLE order_items RENAME COLUMN `order id` TO order_id;',),
 ('ALTER TABLE order_items RENAME COLUMN `order item id` TO order_item_id;',),
 ('ALTER TABLE order_items RENAME COLUMN `price` TO price;',),
 ('ALTER TABLE order_items RENAME COLUMN `product id` TO product_id;',),
 ('ALTER TABLE order_items RENAME COLUMN `seller id` TO seller_id;',),
 ('ALTER TABLE order_items RENAME COLUMN `shipping limit date` TO shipping_limit_date;',),
 ('ALTER TABLE order_payments RENAME COLUMN `order id` TO order_id;',),
 ('ALTER TABLE order_payments RENAME COLUMN `payment installments` TO payment_installments;',),
 ('ALTER TABLE order_payments RENAME COLUMN `payment sequential` TO payment_sequential;',),
 ('ALTER TABLE order_payments RENAME COLUMN `payment type` TO payment_type;',),
 ('ALTER TABLE order_payments RENAME COLUMN `payment value` TO payment_value;',),
 ('ALTER TABLE order_reviews RENAME COLUMN `order id` TO order_id;',),
 ('ALTER TABLE order_reviews RENAME COLUMN `review answer timestamp` TO review_answer_timestamp;',),
 ('ALTER TABLE order_reviews RENAME COLUMN `review comment message` TO review_comment_message;',),
 ('ALTER TABLE order_reviews RENAME COLUMN `review comment title` TO review_comment_title;',),
 ('ALTER TABLE order_reviews RENAME COLUMN `review creation date` TO review_creation_date;',),
 ('ALTER TABLE order_reviews RENAME COLUMN `review id` TO review_id;',),
 ('ALTER TABLE order_reviews RENAME COLUMN `review score` TO review_score;',),
 ('ALTER TABLE orders RENAME COLUMN `customer id` TO customer_id;',),
 ('ALTER TABLE orders RENAME COLUMN `order approved at` TO order_approved_at;',),
 ('ALTER TABLE orders RENAME COLUMN `order delivered carrier date` TO order_delivered_carrier_date;',),
 ('ALTER TABLE orders RENAME COLUMN `order delivered customer date` TO order_delivered_customer_date;',),
 ('ALTER TABLE orders RENAME COLUMN `order estimated delivery date` TO order_estimated_delivery_date;',),
 ('ALTER TABLE orders RENAME COLUMN `order id` TO order_id;',),
 ('ALTER TABLE orders RENAME COLUMN `order purchase timestamp` TO order_purchase_timestamp;',),
 ('ALTER TABLE orders RENAME COLUMN `order status` TO order_status;',),
 ('ALTER TABLE products RENAME COLUMN `product category name` TO product_category_name;',),
 ('ALTER TABLE products RENAME COLUMN `product description lenght` TO product_description_lenght;',),
 ('ALTER TABLE products RENAME COLUMN `product height cm` TO product_height_cm;',),
 ('ALTER TABLE products RENAME COLUMN `product id` TO product_id;',),
 ('ALTER TABLE products RENAME COLUMN `product length cm` TO product_length_cm;',),
 ('ALTER TABLE products RENAME COLUMN `product name lenght` TO product_name_lenght;',),
 ('ALTER TABLE products RENAME COLUMN `product photos qty` TO product_photos_qty;',),
 ('ALTER TABLE products RENAME COLUMN `product weight g` TO product_weight_g;',),
 ('ALTER TABLE products RENAME COLUMN `product width cm` TO product_width_cm;',),
 ('ALTER TABLE sellers RENAME COLUMN `seller city` T

In [ ]:
%%sql

ALTER TABLE category_translation RENAME COLUMN `product_category_name` TO product_category_name;
ALTER TABLE category_translation RENAME COLUMN `product_category_name_english` TO product_category_name_english;
ALTER TABLE customers RENAME COLUMN `customer_city` TO customer_city;
ALTER TABLE customers RENAME COLUMN `customer_id` TO customer_id;
ALTER TABLE customers RENAME COLUMN `customer_state` TO customer_state;
ALTER TABLE customers RENAME COLUMN `customer_unique_id` TO customer_unique_id;
ALTER TABLE customers RENAME COLUMN `customer_zip_code_prefix` TO customer_zip_code_prefix;
ALTER TABLE geolocation RENAME COLUMN `geolocation city` TO geolocation_city;
ALTER TABLE geolocation RENAME COLUMN `geolocation state` TO geolocation_state;
ALTER TABLE geolocation RENAME COLUMN `geolocation zip code prefix` TO geolocation_zip_code_prefix;
ALTER TABLE order_items RENAME COLUMN `freight value` TO freight_value;
ALTER TABLE order_items RENAME COLUMN `order id` TO order_id;
ALTER TABLE order_items RENAME COLUMN `order item id` TO order_item_id;
ALTER TABLE order_items RENAME COLUMN `price` TO price;
ALTER TABLE order_items RENAME COLUMN `product id` TO product_id;
ALTER TABLE order_items RENAME COLUMN `seller id` TO seller_id;
ALTER TABLE order_items RENAME COLUMN `shipping limit date` TO shipping_limit_date;
ALTER TABLE order_payments RENAME COLUMN `order id` TO order_id;
ALTER TABLE order_payments RENAME COLUMN `payment installments` TO payment_installments;
ALTER TABLE order_payments RENAME COLUMN `payment sequential` TO payment_sequential;
ALTER TABLE order_payments RENAME COLUMN `payment type` TO payment_type;
ALTER TABLE order_payments RENAME COLUMN `payment value` TO payment_value;
ALTER TABLE order_reviews RENAME COLUMN `order id` TO order_id;
ALTER TABLE order_reviews RENAME COLUMN `review answer timestamp` TO review_answer_timestamp;
ALTER TABLE order_reviews RENAME COLUMN `review comment message` TO review_comment_message;
ALTER TABLE order_reviews RENAME COLUMN `review comment title` TO review_comment_title;
ALTER TABLE order_reviews RENAME COLUMN `review creation date` TO review_creation_date;
ALTER TABLE order_reviews RENAME COLUMN `review id` TO review_id;
ALTER TABLE order_reviews RENAME COLUMN `review score` TO review_score;
ALTER TABLE orders RENAME COLUMN `customer id` TO customer_id;
ALTER TABLE orders RENAME COLUMN `order approved at` TO order_approved_at;
ALTER TABLE orders RENAME COLUMN `order delivered carrier date` TO order_delivered_carrier_date;
ALTER TABLE orders RENAME COLUMN `order delivered customer date` TO order_delivered_customer_date;
ALTER TABLE orders RENAME COLUMN `order estimated delivery date` TO order_estimated_delivery_date;
ALTER TABLE orders RENAME COLUMN `order id` TO order_id;
ALTER TABLE orders RENAME COLUMN `order purchase timestamp` TO order_purchase_timestamp;
ALTER TABLE orders RENAME COLUMN `order status` TO order_status;
ALTER TABLE products RENAME COLUMN `product category name` TO product_category_name;
ALTER TABLE products RENAME COLUMN `product description lenght` TO product_description_lenght;
ALTER TABLE products RENAME COLUMN `product height cm` TO product_height_cm;
ALTER TABLE products RENAME COLUMN `product id` TO product_id;
ALTER TABLE products RENAME COLUMN `product length cm` TO product_length_cm;
ALTER TABLE products RENAME COLUMN `product name lenght` TO product_name_lenght;
ALTER TABLE products RENAME COLUMN `product photos qty` TO product_photos_qty;
ALTER TABLE products RENAME COLUMN `product weight g` TO product_weight_g;
ALTER TABLE products RENAME COLUMN `product width cm` TO product_width_cm;
ALTER TABLE sellers RENAME COLUMN `seller city` TO seller_city;
ALTER TABLE sellers RENAME COLUMN `seller id` TO seller_id;
ALTER TABLE sellers RENAME COLUMN `seller state` TO seller_state;
ALTER TABLE sellers RENAME COLUMN `seller zip code prefix` TO seller_zip_code_prefix;

## Check if the names are correctly changed 

In [22]:
%%sql

SELECT *

FROM
    orders
LIMIT
    5;

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('e481f51cbdc54678b7cc49136f2d6af7', '9ef432eb6251297304e76186b10a928d', 'delivered', '2017-10-02 10:56:33', '2017-10-02 11:07:15', '2017-10-04 19:55:00', '2017-10-10 21:25:13', '2017-10-18 00:00:00'),
 ('53cdb2fc8bc7dce0b6741e2150273451', 'b0830fb4747a6c6d20dea0b8c802d7ef', 'delivered', '2018-07-24 20:41:37', '2018-07-26 03:24:27', '2018-07-26 14:31:00', '2018-08-07 15:27:45', '2018-08-13 00:00:00'),
 ('47770eb9100c2d0c44946d9cf07ec65d', '41ce2a54c0b03bf3443c3d931a367089', 'delivered', '2018-08-08 08:38:49', '2018-08-08 08:55:23', '2018-08-08 13:50:00', '2018-08-17 18:06:29', '2018-09-04 00:00:00'),
 ('949d5b44dbf5de918fe9c16f97b45f8a', 'f88197465ea7920adcdbec7375364d82', 'delivered', '2017-11-18 19:28:06', '2017-11-18 19:45:59', '2017-11-22 13:39:59', '2017-12-02 00:28:42', '2017-12-15 00:00:00'),
 ('ad21c59c0840e6cb83a9ceb5573f8159', '8ab97904e6daea8866dbdbc4fb7aad2c', 'delivered', '2018-02-13 21:18:39', '2018-02-13 22:20:29', '2018-02-14 19:46:34', '2018-02-16 18:17:02', '2018-02-26 00:00:00')]

In [24]:
%%sql

-- What makes the customer unique id column include duplicates?

SELECT
    *
FROM
    customers
WHERE customer_unique_id in (SELECT customer_unique_id FROM customers group by customer_unique_id having count(customer_unique_id) > 5)
ORDER BY customer_unique_id

LIMIT 25;

 * mysql+pymysql://root:***@localhost:3306/brazil
25 rows affected.


[('a969b9f8d923bc7fd97b578f7c499194', '12f5d6e1cbf93dafd9dcc19095df0b3d', 82200, 'curitiba', 'PR'),
 ('ed1793d2d1e4175d5846ce7ebb4a01f5', '12f5d6e1cbf93dafd9dcc19095df0b3d', 82200, 'curitiba', 'PR'),
 ('27129ce1ebca4ffa23a049c51cc8ec95', '12f5d6e1cbf93dafd9dcc19095df0b3d', 82200, 'curitiba', 'PR'),
 ('588047d7101d88c333691e47659d7099', '12f5d6e1cbf93dafd9dcc19095df0b3d', 82200, 'curitiba', 'PR'),
 ('6152d0774bbbf74f7140541c0569dafa', '12f5d6e1cbf93dafd9dcc19095df0b3d', 82200, 'curitiba', 'PR'),
 ('c0352e94059e3e5a714c9ad0c8306a54', '12f5d6e1cbf93dafd9dcc19095df0b3d', 82200, 'curitiba', 'PR'),
 ('53e16f83735cfe02bf83ef6481ca241e', '1b6c7548a2a1f9037c1fd3ddfed95f33', 38301, 'ituiutaba', 'MG'),
 ('95b9cacc4bc81c3a0a33967bcde6fcd8', '1b6c7548a2a1f9037c1fd3ddfed95f33', 38301, 'ituiutaba', 'MG'),
 ('41a16488effb86580d7a89f8aa19a03b', '1b6c7548a2a1f9037c1fd3ddfed95f33', 38301, 'ituiutaba', 'MG'),
 ('c2ff01bd4fa72f9487ee397f6c5989d2', '1b6c7548a2a1f9037c1fd3ddfed95f33', 38301, 'ituiutaba', 'MG'),
 ('fff7466a253c0e59499ea943462c10f9', '1b6c7548a2a1f9037c1fd3ddfed95f33', 38301, 'ituiutaba', 'MG'),
 ('9a4578c9dd9dce6517d583cb5eca0276', '1b6c7548a2a1f9037c1fd3ddfed95f33', 38301, 'ituiutaba', 'MG'),
 ('ee5c2e36a79081c2440e61990ec2f3af', '1b6c7548a2a1f9037c1fd3ddfed95f33', 38301, 'ituiutaba', 'MG'),
 ('f37a14140ac3a8efbb4d2eab6585edb7', '3e43e6105506432c953e165fb2acf44c', 11701, 'praia grande', 'SP'),
 ('a4413edae691fd0d28bd04b8a38b259d', '3e43e6105506432c953e165fb2acf44c', 11700, 'praia grande', 'SP'),
 ('1815fdaaae692479110428598596b5e5', '3e43e6105506432c953e165fb2acf44c', 11701, 'praia grande', 'SP'),
 ('06ec7aed3161859018db02f695b6600d', '3e43e6105506432c953e165fb2acf44c', 11701, 'praia grande', 'SP'),
 ('6b92dda6e324283dbe4969c443e3bf1a', '3e43e6105506432c953e165fb2acf44c', 11701, 'praia grande', 'SP'),
 ('e68e6423401e85c138229b23d4bf4761', '3e43e6105506432c953e165fb2acf44c', 11700, 'praia grande', 'SP'),
 ('b70f8552b91ef49129519206966e2472', '3e43e6105506432c953e165fb2acf44c', 11700, 'praia grande', 'SP'),
 ('b63c9cfc5dcb06ed307ca95957fa216c', '3e43e6105506432c953e165fb2acf44c', 11704, 'praia grande', 'SP'),
 ('2bbd32d4ef14893d2d8c1a0df08403cf', '3e43e6105506432c953e165fb2acf44c', 11701, 'praia grande', 'SP'),
 ('aa4178995eb97c7a31f3133c5b244651', '47c1a3033b8b77b3ab6e109eb4d5fdf3', 6634, 'jandira', 'SP'),
 ('880b4294c02c693bb9882858758995d5', '47c1a3033b8b77b3ab6e109eb4d5fdf3', 6624, 'jandira', 'SP'),
 ('57802b83580aed851828868d4f7a4917', '47c1a3033b8b77b3ab6e109eb4d5fdf3', 6634, 'jandira', 'SP')]

AS we see the real customer ID is the column named (customer_unique_id), We should invistigate this further.

After some online invistigation I found out that OLIST system generates a unique id for each customer right after signing up so it's permanent, While generates a random ID every time a cutomer places an order which is stored in customer id.
We can check this by comparing number of unique orders id and our customer id column from the customers table.

In [13]:
%sql SELECT COUNT(distinct customer_id) from customers;

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(99441,)]

In [25]:
%sql SELECT COUNT(distinct order_id) from orders;

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(99441,)]

In [26]:
%sql SELECT * FROM orders limit 5

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('e481f51cbdc54678b7cc49136f2d6af7', '9ef432eb6251297304e76186b10a928d', 'delivered', '2017-10-02 10:56:33', '2017-10-02 11:07:15', '2017-10-04 19:55:00', '2017-10-10 21:25:13', '2017-10-18 00:00:00'),
 ('53cdb2fc8bc7dce0b6741e2150273451', 'b0830fb4747a6c6d20dea0b8c802d7ef', 'delivered', '2018-07-24 20:41:37', '2018-07-26 03:24:27', '2018-07-26 14:31:00', '2018-08-07 15:27:45', '2018-08-13 00:00:00'),
 ('47770eb9100c2d0c44946d9cf07ec65d', '41ce2a54c0b03bf3443c3d931a367089', 'delivered', '2018-08-08 08:38:49', '2018-08-08 08:55:23', '2018-08-08 13:50:00', '2018-08-17 18:06:29', '2018-09-04 00:00:00'),
 ('949d5b44dbf5de918fe9c16f97b45f8a', 'f88197465ea7920adcdbec7375364d82', 'delivered', '2017-11-18 19:28:06', '2017-11-18 19:45:59', '2017-11-22 13:39:59', '2017-12-02 00:28:42', '2017-12-15 00:00:00'),
 ('ad21c59c0840e6cb83a9ceb5573f8159', '8ab97904e6daea8866dbdbc4fb7aad2c', 'delivered', '2018-02-13 21:18:39', '2018-02-13 22:20:29', '2018-02-14 19:46:34', '2018-02-16 18:17:02', '2018-02-26 00:00:00')]

In [27]:
%sql SELECT COUNT(distinct customer_id) from orders;

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(99441,)]

In [32]:
%%sql 

SELECT * from Orders limit 5;

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('e481f51cbdc54678b7cc49136f2d6af7', '9ef432eb6251297304e76186b10a928d', 'delivered', '2017-10-02 10:56:33', '2017-10-02 11:07:15', '2017-10-04 19:55:00', '2017-10-10 21:25:13', '2017-10-18 00:00:00'),
 ('53cdb2fc8bc7dce0b6741e2150273451', 'b0830fb4747a6c6d20dea0b8c802d7ef', 'delivered', '2018-07-24 20:41:37', '2018-07-26 03:24:27', '2018-07-26 14:31:00', '2018-08-07 15:27:45', '2018-08-13 00:00:00'),
 ('47770eb9100c2d0c44946d9cf07ec65d', '41ce2a54c0b03bf3443c3d931a367089', 'delivered', '2018-08-08 08:38:49', '2018-08-08 08:55:23', '2018-08-08 13:50:00', '2018-08-17 18:06:29', '2018-09-04 00:00:00'),
 ('949d5b44dbf5de918fe9c16f97b45f8a', 'f88197465ea7920adcdbec7375364d82', 'delivered', '2017-11-18 19:28:06', '2017-11-18 19:45:59', '2017-11-22 13:39:59', '2017-12-02 00:28:42', '2017-12-15 00:00:00'),
 ('ad21c59c0840e6cb83a9ceb5573f8159', '8ab97904e6daea8866dbdbc4fb7aad2c', 'delivered', '2018-02-13 21:18:39', '2018-02-13 22:20:29', '2018-02-14 19:46:34', '2018-02-16 18:17:02', '2018-02-26 00:00:00')]

Looking for the previous resultset and checking the schema picture I provided above now I know that Customer.customer_id
is used only to connect etween customers and orders tables. I thought about dropping orders.oreder_id and using orders.customer_id to connect the orders table to other tables too But that's unneccesary effort. I'll keep everything as it is.
<font color='green'> <b> But keep in mind that any customer aggragation will be done using customer.customer_unique_id.<b/><font/>

_______________________________________________________________________________________________________________________________
# <FONT color = 'red'>Check the integrity of columns nominated to be primary keys<FONT/>

_______________________________________________________________________________________________________________________________

## <FONT color = 'green'>1- Customers table<FONT/>

In [29]:
%sql describe customers

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('customer_id', 'varchar(32)', 'NO', '', None, ''),
 ('customer_unique_id', 'varchar(32)', 'NO', '', None, ''),
 ('customer_zip_code_prefix', 'int', 'NO', '', None, ''),
 ('customer_city', 'varchar(32)', 'NO', '', None, ''),
 ('customer_state', 'varchar(2)', 'NO', '', None, '')]

In [30]:
%sql SELECT COUNT(customer_id), COUNT(distinct customer_id) FROM customers;


 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(99441, 99441)]

In [ ]:
# Now I'm creating a primary key for the custoers table
%sql Alter table customers ADD primary key (customer_id);

_______________________________________________________________________________________________________________________________
## <FONT color = 'green'>2- Orders table<FONT/>

In [33]:
%sql Describe orders; -- Checking for keys

 * mysql+pymysql://root:***@localhost:3306/brazil
8 rows affected.


[('order_id', 'varchar(32)', 'NO', '', None, ''),
 ('customer_id', 'varchar(32)', 'NO', '', None, ''),
 ('order_status', 'varchar(11)', 'NO', '', None, ''),
 ('order_purchase_timestamp', 'varchar(19)', 'NO', '', None, ''),
 ('order_approved_at', 'varchar(19)', 'YES', '', None, ''),
 ('order_delivered_carrier_date', 'varchar(19)', 'YES', '', None, ''),
 ('order_delivered_customer_date', 'varchar(19)', 'YES', '', None, ''),
 ('order_estimated_delivery_date', 'varchar(19)', 'NO', '', None, '')]

In [34]:
%sql SELECT COUNT(order_id), COUNT(distinct order_id) FROM orders;

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(99441, 99441)]

In [ ]:
# Now I'm creating a primary key for the Orders table
%sql Alter table orders ADD primary key (order_id);

_______________________________________________________________________________________________________________________________
## <FONT color = 'green'>3- Geo table<FONT/>

In [37]:
%sql Describe geolocation;

 * mysql+pymysql://root:***@localhost:3306/brazil
3 rows affected.


[('geolocation_zip_code_prefix', 'int', 'NO', '', None, ''),
 ('geolocation_city', 'varchar(38)', 'NO', '', None, ''),
 ('geolocation_state', 'varchar(2)', 'NO', '', None, '')]

In [38]:
%%sql 
SELECT 
    Count(geolocation_zip_code_prefix),
    COUNT(distinct geolocation_zip_code_prefix), 
    COUNT(distinct geolocation_zip_code_prefix, geolocation_city, geolocation_state) 
FROM 
    geolocation;


 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(1000163, 19015, 19617)]

Wow, That's a huge amount of duplicates. As you see there are 19617 unique compinations of all columns but the table has over 
<font color='red'> one million </font> row.
Lets remove those extra rows. We'll do that by creating another table containing the unique compinations only.

In [ ]:
%%sql

CREATE TABLE geo AS(
SELECT
    geolocation_zip_code_prefix, 
    MAX(geolocation_city), 
    MAX(geolocation_state)
    
FROM 
    geolocation
    
GROUP BY 
    geolocation_zip_code_prefix
);

In [40]:
%%sql

SELECT 
    COUNT(*)
    
FROM
    geo;
    

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(19015,)]

In [ ]:
#Let's drop the old table
%sql DROP TABLE geolocation;

In [42]:
%sql Describe geo;

 * mysql+pymysql://root:***@localhost:3306/brazil
3 rows affected.


[('geolocation_zip_code_prefix', 'int', 'NO', '', None, ''),
 ('MAX(geolocation_city)', 'varchar(38)', 'YES', '', None, ''),
 ('MAX(geolocation_state)', 'varchar(2)', 'YES', '', None, '')]

In [ ]:
%%sql 
ALTER TABLE geo
RENAME TO geolocation;

In [ ]:
# Now I'm creating a primary key for the geolocation table
%sql Alter table geolocation ADD primary key (geolocation_zip_code_prefix);

_______________________________________________________________________________________________________________________________
## <FONT color = 'green'>4- Order reviews table<FONT/>

In [66]:
%sql SELECT * FROM order_reviews LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('7bc2406110b926393aa56f80a40eba40', '73fc7af87114b39712e6da79b0a377eb', 4, None, None, '2018-01-18 00:00:00', '2018-01-18 21:46:59'),
 ('80e641a11e56f04c1ad469d5645fdfde', 'a548910a1c6147796b98fdf73dbeba33', 5, None, None, '2018-03-10 00:00:00', '2018-03-11 03:05:13'),
 ('228ce5500dc1d8e020d8d1322874b6f0', 'f9e4b658b201a9f2ecdecbb34bed034b', 5, None, None, '2018-02-17 00:00:00', '2018-02-18 14:36:24'),
 ('e64fb393e7b32834bb789ff8bb30750e', '658677c97b385a9be170737859d3511b', 5, None, 'Recebi bem antes do prazo estipulado.', '2017-04-21 00:00:00', '2017-04-21 22:02:06'),
 ('f7c4243c7fe1938f181bec41a392bdeb', '8e6bfb81e283fa7e4f11123a3fb894f1', 5, None, 'Parabéns lojas lannister adorei comprar pela Internet seguro e prático Parabéns a todos feliz Páscoa', '2018-03-01 00:00:00', '2018-03-02 10:26:53')]

In [5]:
%%sql
SELECT
    count(review_id), count(Distinct review_id)
FROM
    order_reviews;

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(99224, 98410)]

In [6]:
%%sql
-- CHECK for review_id present more than once in the table
SELECT
    *
FROM
    order_reviews
    
WHERE 
    review_id IN (
                    SELECT 
                        review_id
                    FROM
                        order_reviews
                    GROUP BY review_id
                    HAVING COUNT(review_id) > 1
                    )
ORDER BY 
    review_id;

 * mysql+pymysql://root:***@localhost:3306/brazil
1603 rows affected.


[('00130cbe1f9d422698c812ed8ded1919', '04a28263e085d399c97ae49e0b477efa', 1, None, 'O cartucho "original HP" 60XL não é reconhecido pela impressora, consequentemente não funcionou. Além de ter chegado com atraso de mais de 15 dias do previsto. Preciso que seja trocado.', '2018-03-07 00:00:00', '2018-03-20 18:08:23'),
 ('00130cbe1f9d422698c812ed8ded1919', 'dfcdfc43867d1c1381bfaf62d6b9c195', 1, None, 'O cartucho "original HP" 60XL não é reconhecido pela impressora, consequentemente não funcionou. Além de ter chegado com atraso de mais de 15 dias do previsto. Preciso que seja trocado.', '2018-03-07 00:00:00', '2018-03-20 18:08:23'),
 ('0115633a9c298b6a98bcbe4eee75345f', '0c9850b2c179c1ef60d2855e2751d1fa', 5, None, None, '2017-09-21 00:00:00', '2017-09-26 03:27:47'),
 ('0115633a9c298b6a98bcbe4eee75345f', '78a4201f58af3463bdab842eea4bc801', 5, None, None, '2017-09-21 00:00:00', '2017-09-26 03:27:47'),
 ('0174caf0ee5964646040cd94e15ac95e', 'f93a732712407c02dce5dd5088d0f47b', 1, None, 'Produto entregue dentro de embalagem do fornecedor sem os parafusos de fixação das partes.', '2018-03-07 00:00:00', '2018-03-08 03:00:53'),
 ('0174caf0ee5964646040cd94e15ac95e', '74db91e33b4e1fd865356c89a61abf1f', 1, None, 'Produto entregue dentro de embalagem do fornecedor sem os parafusos de fixação das partes.', '2018-03-07 00:00:00', '2018-03-08 03:00:53'),
 ('017808d29fd1f942d97e50184dfb4c13', '8daaa9e99d60fbba579cc1c3e3bfae01', 5, None, None, '2018-03-02 00:00:00', '2018-03-05 01:43:30'),
 ('017808d29fd1f942d97e50184dfb4c13', 'b1461c8882153b5fe68307c46a506e39', 5, None, None, '2018-03-02 00:00:00', '2018-03-05 01:43:30'),
 ('0254bd905dc677a6078990aad3331a36', '331b367bdd766f3d1cf518777317b5d9', 1, None, 'O pedido consta de 2 produtos e até agora recebi apenas 1 produto, e o que me preocupa é que o status aparece como entregue.\nSolicito providências', '2017-09-09 00:00:00', '2017-09-13 09:52:44'),
 ('0254bd905dc677a6078990aad3331a36', '5bf226cf882c5bf4247f89a97c86f273', 1, None, 'O pedido consta de 2 produtos e até agora recebi apenas 1 produto, e o que me preocupa é que o status aparece como entregue.\nSolicito providências', '2017-09-09 00:00:00', '2017-09-13 09:52:44'),
 ('0288d42bef3dfe36930740c9588a570f', '33d8795f04dd631f3480d7aaf90da3dc', 3, None, None, '2017-07-29 00:00:00', '2017-07-31 17:11:04'),
 ('0288d42bef3dfe36930740c9588a570f', 'f889a5a0b44adc29c5465b99395ac3c1', 3, None, None, '2017-07-29 00:00:00', '2017-07-31 17:11:04'),
 ('02aa7f5f75e964e3c7efa59a1f515281', 'db92613c074f00e53066388d48ad7512', 3, None, None, '2017-07-27 00:00:00', '2017-07-29 15:09:50'),
 ('02aa7f5f75e964e3c7efa59a1f515281', 'd8e17d5f7dacf0970d316e7c03e741e8', 3, None, None, '2017-07-27 00:00:00', '2017-07-29 15:09:50'),
 ('034528cb00b6bc981847acafbdf2ae0b', 'e1e8e3bca903de27e9a1c72b5a5795e0', 1, None, None, '2017-05-14 00:00:00', '2017-05-15 02:17:38'),
 ('034528cb00b6bc981847acafbdf2ae0b', '3d3742a96f24a8fe4e2e57628807e476', 1, None, None, '2017-05-14 00:00:00', '2017-05-15 02:17:38'),
 ('03a6a25db577d0689440933055111897', '2acfdc5131ff2cf4433e668454c9784c', 5, None, 'Muito Bom! Gostei Bastante! Tecido Ótimo! Atendeu Minhas Espectativas! 👏👏👏 Super Recomendo! 👍', '2017-12-15 00:00:00', '2017-12-16 01:32:18'),
 ('03a6a25db577d0689440933055111897', '3fde8b7313af6b37b84b5c7594d7add0', 5, None, 'Muito Bom! Gostei Bastante! Tecido Ótimo! Atendeu Minhas Espectativas! 👏👏👏 Super Recomendo! 👍', '2017-12-15 00:00:00', '2017-12-16 01:32:18'),
 ('0467560f511c516ddaa54a60edb0c291', '55b2e390d5d80ada31ad1b795ebeb087', 5, None, None, '2017-02-09 00:00:00', '2017-02-14 15:58:44'),
 ('0467560f511c516ddaa54a60edb0c291', '0c995611a99f81268d859184a416f1db', 5, None, None, '2017-02-09 00:00:00', '2017-02-14 15:58:44'),
 ('047fd109ced39e02296f6aeb74f6a6f1', '236e6ec6171c1870d4bcf4ccfad87f49', 4, None, None, '2018-07-31 00:00:00', '2018-08-02 15:11:49'),
 ('047fd109ced39e02296f6aeb74f6a6f1', 'a89abace0dcc01eeb267a9660b5ac126', 4, None, None, '2018-07-31 00:00:00', '2018-08-02 15:11:

In [7]:
%%sql
-- Check for certain values for more investigation
SELECT *
FROM order_items
WHERE order_id in ('d75cb3755738c4ae466303358f97bc55', '0068c109948b9a1dfb8530d1978acef3');

 * mysql+pymysql://root:***@localhost:3306/brazil
3 rows affected.


[('0068c109948b9a1dfb8530d1978acef3', 1, '525f8fbde09c9bf777025c3141fbbc6e', '8b321bb669392f5163d04c59e235e066', '2018-01-30 02:53:04', Decimal('35.90'), Decimal('15.79')),
 ('0068c109948b9a1dfb8530d1978acef3', 2, '525f8fbde09c9bf777025c3141fbbc6e', '8b321bb669392f5163d04c59e235e066', '2018-01-30 02:53:04', Decimal('35.90'), Decimal('15.79')),
 ('d75cb3755738c4ae466303358f97bc55', 1, '474bb6b54fc608ca71059a6c4f7ecda3', '8b321bb669392f5163d04c59e235e066', '2018-01-30 02:53:17', Decimal('21.90'), Decimal('15.79'))]

Notice that some reviews are identical for differnet orders, This needs more investigation.

In [ ]:
# Now I'm creating a primary key for the order_reviews table
%sql Alter table order_reviews ADD primary key (review_id, order_id);

_______________________________________________________________________________________________________________________________
## <FONT color = 'green'>5- Order items table<FONT/>

In [9]:
%sql SELECT * FROM order_items LIMIT 5

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('00010242fe8c5a6d1ba2dd792cb16214', 1, '4244733e06e7ecb4970a6e2683c13e61', '48436dade18ac8b2bce089ec2a041202', '2017-09-19 09:45:35', Decimal('58.90'), Decimal('13.29')),
 ('00018f77f2f0320c557190d7a144bdd3', 1, 'e5f2d52b802189ee658865ca93d83a8f', 'dd7ddc04e1b6c2c614352b383efe2d36', '2017-05-03 11:05:13', Decimal('239.90'), Decimal('19.93')),
 ('000229ec398224ef6ca0657da4fc703e', 1, 'c777355d18b72b67abbeef9df44fd0fd', '5b51032eddd242adc84c38acab88f23d', '2018-01-18 14:48:30', Decimal('199.00'), Decimal('17.87')),
 ('00024acbcdf0a6daa1e931b038114c75', 1, '7634da152a4610f1595efa32f14722fc', '9d7a1d34a5052409006425275ba1c2b4', '2018-08-15 10:10:18', Decimal('12.99'), Decimal('12.79')),
 ('00042b26cf59d7ce69dfabb4e55b4fd9', 1, 'ac6c3623068f30de03045865e4e10089', 'df560393f3a51e74553ab94004ba5c87', '2017-02-13 13:57:51', Decimal('199.90'), Decimal('18.14'))]

In [10]:
%sql SELECT COUNT(order_id), COUNT(DISTINCT order_id, product_id), COUNT(DISTINCT order_id, product_id, seller_id) FROM order_items

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(112650, 102425, 102425)]

From the previous result set it's clear that we have duplicates in this table.
After removing duplicates we'll assign the compination of order_id, product id as a primary key.

In [11]:
%%sql
-- Check if the join with distinct values return the right row count 102425 as in the previous query.
SELECT COUNT(*)
FROM
(SELECT 
    order_id,
    order_item_id,
    product_id,
    seller_id,
    shipping_limit_date,
    price,
    freight_value,
    ROW_NUMBER() OVER (PARTITION BY order_id, product_id ORDER BY order_id, product_id) AS rn
    
FROM 
    order_items) as desired_table
WHERE rn = 1

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(102425,)]

In [ ]:
%%sql
-- The code is returning the desired output now we will create that contains the rows with unique (order_id, product_id)

Create table oi as(
SELECT *
FROM
(SELECT 
    order_id,
    order_item_id,
    product_id,
    seller_id,
    shipping_limit_date,
    price,
    freight_value,
    ROW_NUMBER() OVER (PARTITION BY order_id, product_id ORDER BY order_id, product_id) AS rn
    
FROM 
    order_items) as desired_table
WHERE rn = 1
)

 Now remove the old table and rename the new one with the desired name(oeder_items)

In [ ]:
%sql DROP TABLE order_items;

In [ ]:
%sql ALTER TABLE oi RENAME order_items

In [ ]:
# Now I'm creating a primary key for the order_items table
%sql Alter table order_items ADD primary key (order_id, product_id);

_______________________________________________________________________________________________________________________________
## <FONT color = 'green'>6- Products table<FONT/>

In [17]:
%sql SELECT * FROM products limit 5;

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('1e9e8ef04dbcff4541ed26657ea517e5', 'perfumaria', '40', '287', '1', '225', '16', '10', '14'),
 ('3aa071139cb16b67ca9e5dea641aaa2f', 'artes', '44', '276', '1', '1000', '30', '18', '20'),
 ('96bd76ec8810374ed1b65e291975717f', 'esporte_lazer', '46', '250', '1', '154', '18', '9', '15'),
 ('cef67bcfe19066a932b7673e239eb23d', 'bebes', '27', '261', '1', '371', '26', '4', '26'),
 ('9dc1a7de274444849c219cff195d0b71', 'utilidades_domesticas', '37', '402', '4', '625', '20', '17', '13')]

In [18]:
%%sql

SELECT
    COUNT(product_id),
    COUNT(DISTINCT product_id) 

FROM products;

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(32951, 32951)]

As you see there is no duplicates the column is ready to be assigned as a primary key.

In [ ]:
# Now I'm creating a primary key for the order_items table
%sql Alter table products ADD primary key (product_id);

_______________________________________________________________________________________________________________________________
## <FONT color = 'green'>7- Order payments table<FONT/>

In [19]:
%sql SELECT * FROM order_payments limit 5

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('b81ef226f3fe1789b1e8b2acac839d17', 1, 'credit_card', 8, Decimal('99.33')),
 ('a9810da82917af2d9aefd1278f1dcfa0', 1, 'credit_card', 1, Decimal('24.39')),
 ('25e8ea4e93396b6fa0d3dd708e76c1bd', 1, 'credit_card', 1, Decimal('65.71')),
 ('ba78997921bbcdc1373bb41e913ab953', 1, 'credit_card', 8, Decimal('107.78')),
 ('42fdf880ba16b47b59251dd489d4441a', 1, 'credit_card', 2, Decimal('128.45'))]

In [20]:
%sql SELECT COUNT(order_id), COUNT( DISTINCT order_id) FROM order_payments

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(103886, 99440)]

So as you see some orders are mentioned more than once, Lets check them.

In [21]:
%%sql


SELECT *

FROM 
    order_payments
    
WHERE order_id in (SELECT 
                       order_id                        
                   FROM 
                       order_payments
                   GROUP BY 
                       order_id
                   HAVING
                      COUNT(order_id) > 1)
ORDER BY order_id
LIMIT 10

 * mysql+pymysql://root:***@localhost:3306/brazil
10 rows affected.


[('0016dfedd97fc2950e388d2971d718c7', 2, 'voucher', 1, Decimal('17.92')),
 ('0016dfedd97fc2950e388d2971d718c7', 1, 'credit_card', 5, Decimal('52.63')),
 ('002f19a65a2ddd70a090297872e6d64e', 1, 'voucher', 1, Decimal('44.11')),
 ('002f19a65a2ddd70a090297872e6d64e', 2, 'voucher', 1, Decimal('33.18')),
 ('0071ee2429bc1efdc43aa3e073a5290e', 1, 'voucher', 1, Decimal('100.00')),
 ('0071ee2429bc1efdc43aa3e073a5290e', 2, 'voucher', 1, Decimal('92.44')),
 ('009ac365164f8e06f59d18a08045f6c4', 5, 'voucher', 1, Decimal('8.75')),
 ('009ac365164f8e06f59d18a08045f6c4', 1, 'credit_card', 1, Decimal('0.88')),
 ('009ac365164f8e06f59d18a08045f6c4', 2, 'voucher', 1, Decimal('4.50')),
 ('009ac365164f8e06f59d18a08045f6c4', 6, 'voucher', 1, Decimal('4.17'))]

In [22]:
%sql SELECT COUNT(*), COUNT( DISTINCT order_id, payment_sequential) FROM order_payments

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(103886, 103886)]

As you see it's normal that an order in mentioned more than once, other attributes changes.
Which indicates that the order maybe paid with different methods and with many payments. 

In [ ]:
# Now I'm creating a primary key for the order_payments table
%sql Alter table order_payments ADD primary key (order_id, payment_sequential);

_______________________________________________________________________________________________________________________________
## <FONT color = 'green'>8- Sellers table<FONT/>

In [24]:
%sql SELECT * FROM sellers limit 5

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('3442f8959a84dea7ee197c632cb2df15', 13023, 'campinas', 'SP'),
 ('d1b65fc7debc3361ea86b5f14c68d2e2', 13844, 'mogi guacu', 'SP'),
 ('ce3ad9de960102d0677a81f5d0bb7b2d', 20031, 'rio de janeiro', 'RJ'),
 ('c0f3eea2e14555b6faeea3dd58c1b1c3', 4195, 'sao paulo', 'SP'),
 ('51a04a8a6bdcb23deccc82b0b80742cf', 12914, 'braganca paulista', 'SP')]

In [25]:
%sql SELECT COUNT(seller_id), COUNT(DISTINCT seller_id) FROM sellers;

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(3095, 3095)]

In [ ]:
# Now I'm creating a primary key for the sellers table
%sql Alter table sellers ADD primary key (seller_id);

In [27]:
%sql describe sellers

 * mysql+pymysql://root:***@localhost:3306/brazil
4 rows affected.


[('seller_id', 'varchar(32)', 'NO', 'PRI', None, ''),
 ('seller_zip_code_prefix', 'int', 'NO', '', None, ''),
 ('seller_city', 'varchar(40)', 'NO', '', None, ''),
 ('seller_state', 'varchar(2)', 'NO', '', None, '')]

_______________________________________________________________________________________________________________________________
## <FONT color = 'green'>9- sellersproduct_category_name_translation table<FONT/>

In [28]:
%sql SELECT * FROM category_translation LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/brazil
10 rows affected.


[('agro_industria_e_comercio', 'agro_industry_and_commerce'),
 ('alimentos', 'food'),
 ('alimentos_bebidas', 'food_drink'),
 ('artes', 'art'),
 ('artes_e_artesanato', 'arts_and_craftmanship'),
 ('artigos_de_festas', 'party_supplies'),
 ('artigos_de_natal', 'christmas_supplies'),
 ('audio', 'audio'),
 ('automotivo', 'auto'),
 ('bebes', 'baby')]

In [29]:
%sql SELECT COUNT(product_category_name), COUNT(DISTINCT product_category_name) FROM category_translation;

 * mysql+pymysql://root:***@localhost:3306/brazil
1 rows affected.


[(71, 71)]

In [ ]:
# Now I'm creating a primary key for the category_translation table
%sql Alter table category_translation ADD primary key (product_category_name);

In [32]:
%sql describe category_translation

 * mysql+pymysql://root:***@localhost:3306/brazil
2 rows affected.


[('product_category_name', 'varchar(46)', 'NO', 'PRI', None, ''),
 ('product_category_name_english', 'varchar(39)', 'NO', '', None, '')]

_______________________________________________________________________________________________________________________________

_______________________________________________________________________________________________________________________________
# <FONT color = 'red'>Check the integrity of columns nominated to be foreign keys and creating relationships between tables<FONT/>

In [33]:
%sql SHOW TABLES

 * mysql+pymysql://root:***@localhost:3306/brazil
9 rows affected.


[('category_translation',),
 ('customers',),
 ('geolocation',),
 ('order_items',),
 ('order_payments',),
 ('order_reviews',),
 ('orders',),
 ('products',),
 ('sellers',)]

### <font color='Indigo'>Desired Schema <font/>

<img src="https://i.imgur.com/HRhd2Y0.png" alt="Desired Schema"/>


_______________________________________________________________________________________________________________________________

## <FONT color='green'>Customers - Geolocation using zip_code_prefix</FONT>

In [ ]:
%%sql
ALTER Table geolocation
RENAME COLUMN `MAX(geolocation_city)` TO geolocation_city,
RENAME COLUMN `MAX(geolocation_state)` TO geolocation_state;

In [ ]:
%%sql

ALTER TABLE customers
ADD CONSTRAINT fk_name1
FOREIGN KEY (customer_zip_code_prefix)
REFERENCES geolocation(geolocation_zip_code_prefix)

there are values in the customers table that does not have reference values in geolocation table.
Since customer table hav the all columns of the geo table, I'll add values from customers to geolocation.

In [ ]:

%%sql 
INSERT INTO geolocation 
(
SELECT 
    customer_zip_code_prefix,
    MAX(customer_city) as customer_city,
    MAX(customer_state) as customer_state
FROM 
    CUSTOMERS
WHERE 
    customer_zip_code_prefix NOT IN (SELECT 
                                         geolocation_zip_code_prefix 
                                     FROM
                                         geolocation)
GROUP BY 
    customer_zip_code_prefix  -- use group by to eleminate duplicate values.
    
    
)

In [51]:
%%sql 
SELECT * 
FROM geolocation 
LIMIT 5

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[(1001, 'sao paulo', 'SP'),
 (1002, 'sao paulo', 'SP'),
 (1003, 'sao paulo', 'SP'),
 (1004, 'sao paulo', 'SP'),
 (1005, 'sao paulo', 'SP')]

In [52]:
%%sql
-- Check for values in customer table but missing in geolocation.
SELECT 
    * 
FROM 
    customers 
WHERE 
    customer_zip_code_prefix not in (
                                    SELECT 
                                        geolocation_zip_code_prefix 
                                    FROM 
                                        geolocation)

 * mysql+pymysql://root:***@localhost:3306/brazil
0 rows affected.


[]

In [ ]:
%%sql

ALTER TABLE customers
ADD CONSTRAINT fk_name1
FOREIGN KEY (customer_zip_code_prefix)
REFERENCES geolocation(geolocation_zip_code_prefix)

_______________________________________________________________________________________________________________________________

## <FONT color='green'>Sellers - Geolocation using zip_code_prefix</FONT>



In [54]:
%%sql
SELECT *
FROM 
    sellers
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('0015a82c2db000af6aaaf3ae2ecb0532', 9080, 'santo andre', 'SP'),
 ('001cca7ae9ae17fb1caed9dfb1094831', 29156, 'cariacica', 'ES'),
 ('001e6ad469a905060d959994f1b41e4f', 24754, 'sao goncalo', 'RJ'),
 ('002100f778ceb8431b7a1020ff7ab48f', 14405, 'franca', 'SP'),
 ('003554e2dce176b5555353e4f3555ac8', 74565, 'goiania', 'GO')]

We had the same previous problem about data missing in the geolocation table. We'll add missing data from sellers into geolocation.

In [ ]:
%%sql 
INSERT INTO geolocation 
(
SELECT 
    seller_zip_code_prefix,
    MAX(seller_city) as seller_city,
    MAX(seller_state) as seller_state
FROM 
    SELLERS
WHERE 
    seller_zip_code_prefix NOT IN (SELECT 
                                         geolocation_zip_code_prefix 
                                     FROM
                                         geolocation)
GROUP BY 
    seller_zip_code_prefix  -- use group by to eleminate duplicate values.
    
)



_______________________________________________________________________________________________________________________________

## <FONT color='green'>Customers</FONT> - <FONT color='grey'><FONT color='green'>Geolocation</FONT> </FONT><FONT color='grey'>relationship</FONT>


In [ ]:
%%sql

ALTER TABLE sellers
ADD CONSTRAINT fk_name2
FOREIGN KEY (seller_zip_code_prefix)
REFERENCES geolocation(geolocation_zip_code_prefix);

_______________________________________________________________________________________________________________________________

## <FONT color='green'>Customers - Orders using customer_id</FONT>



In [42]:
%%sql
SELECT * 
FROM orders
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('00010242fe8c5a6d1ba2dd792cb16214', '3ce436f183e68e07877b285a838db11a', 'delivered', '2017-09-13 08:59:02', '2017-09-13 09:45:35', '2017-09-19 18:34:16', '2017-09-20 23:43:48', '2017-09-29 00:00:00'),
 ('00018f77f2f0320c557190d7a144bdd3', 'f6dd3ec061db4e3987629fe6b26e5cce', 'delivered', '2017-04-26 10:53:06', '2017-04-26 11:05:13', '2017-05-04 14:35:00', '2017-05-12 16:04:24', '2017-05-15 00:00:00'),
 ('000229ec398224ef6ca0657da4fc703e', '6489ae5e4333f3693df5ad4372dab6d3', 'delivered', '2018-01-14 14:33:31', '2018-01-14 14:48:30', '2018-01-16 12:36:48', '2018-01-22 13:19:16', '2018-02-05 00:00:00'),
 ('00024acbcdf0a6daa1e931b038114c75', 'd4eb9395c8c0431ee92fce09860c5a06', 'delivered', '2018-08-08 10:00:35', '2018-08-08 10:10:18', '2018-08-10 13:28:00', '2018-08-14 13:32:39', '2018-08-20 00:00:00'),
 ('00042b26cf59d7ce69dfabb4e55b4fd9', '58dbd0b2d70206bf40e62cd34e84d795', 'delivered', '2017-02-04 13:57:51', '2017-02-04 14:10:13', '2017-02-16 09:46:09', '2017-03-01 16:42:31', '2017-03-17 00:00:00')]

In [ ]:
%%sql

ALTER TABLE orders
ADD CONSTRAINT fk_name3
FOREIGN KEY (customer_id)
REFERENCES customers(customer_id);

_______________________________________________________________________________________________________________________________

## <FONT color='green'>Sellers - Order_items using seller_id</FONT>



In [44]:
%%sql

SELECT *
FROM order_items
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('00010242fe8c5a6d1ba2dd792cb16214', 1, '4244733e06e7ecb4970a6e2683c13e61', '48436dade18ac8b2bce089ec2a041202', '2017-09-19 09:45:35', Decimal('58.90'), Decimal('13.29'), 1),
 ('00018f77f2f0320c557190d7a144bdd3', 1, 'e5f2d52b802189ee658865ca93d83a8f', 'dd7ddc04e1b6c2c614352b383efe2d36', '2017-05-03 11:05:13', Decimal('239.90'), Decimal('19.93'), 1),
 ('000229ec398224ef6ca0657da4fc703e', 1, 'c777355d18b72b67abbeef9df44fd0fd', '5b51032eddd242adc84c38acab88f23d', '2018-01-18 14:48:30', Decimal('199.00'), Decimal('17.87'), 1),
 ('00024acbcdf0a6daa1e931b038114c75', 1, '7634da152a4610f1595efa32f14722fc', '9d7a1d34a5052409006425275ba1c2b4', '2018-08-15 10:10:18', Decimal('12.99'), Decimal('12.79'), 1),
 ('00042b26cf59d7ce69dfabb4e55b4fd9', 1, 'ac6c3623068f30de03045865e4e10089', 'df560393f3a51e74553ab94004ba5c87', '2017-02-13 13:57:51', Decimal('199.90'), Decimal('18.14'), 1)]

In [ ]:
%%sql

ALTER TABLE order_items
ADD CONSTRAINT fk_name4
FOREIGN KEY (seller_id)
REFERENCES sellers(seller_id);

_______________________________________________________________________________________________________________________________

## <FONT color='green'>Orders - Order_items using order_id</FONT>



In [46]:
%%sql
SELECT * FROM orders limit 5;

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('00010242fe8c5a6d1ba2dd792cb16214', '3ce436f183e68e07877b285a838db11a', 'delivered', '2017-09-13 08:59:02', '2017-09-13 09:45:35', '2017-09-19 18:34:16', '2017-09-20 23:43:48', '2017-09-29 00:00:00'),
 ('00018f77f2f0320c557190d7a144bdd3', 'f6dd3ec061db4e3987629fe6b26e5cce', 'delivered', '2017-04-26 10:53:06', '2017-04-26 11:05:13', '2017-05-04 14:35:00', '2017-05-12 16:04:24', '2017-05-15 00:00:00'),
 ('000229ec398224ef6ca0657da4fc703e', '6489ae5e4333f3693df5ad4372dab6d3', 'delivered', '2018-01-14 14:33:31', '2018-01-14 14:48:30', '2018-01-16 12:36:48', '2018-01-22 13:19:16', '2018-02-05 00:00:00'),
 ('00024acbcdf0a6daa1e931b038114c75', 'd4eb9395c8c0431ee92fce09860c5a06', 'delivered', '2018-08-08 10:00:35', '2018-08-08 10:10:18', '2018-08-10 13:28:00', '2018-08-14 13:32:39', '2018-08-20 00:00:00'),
 ('00042b26cf59d7ce69dfabb4e55b4fd9', '58dbd0b2d70206bf40e62cd34e84d795', 'delivered', '2017-02-04 13:57:51', '2017-02-04 14:10:13', '2017-02-16 09:46:09', '2017-03-01 16:42:31', '2017-03-17 00:00:00')]

In [ ]:
%%sql

ALTER TABLE order_items
ADD CONSTRAINT fk_name5
FOREIGN KEY (order_id)
REFERENCES orders(order_id);

_______________________________________________________________________________________________________________________________

## <FONT color='green'>Products - Order_items using product_id</FONT>



In [57]:
%%sql
SELECT * FROM products limit 5;

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('1e9e8ef04dbcff4541ed26657ea517e5', 'perfumaria', '40', '287', '1', '225', '16', '10', '14'),
 ('3aa071139cb16b67ca9e5dea641aaa2f', 'artes', '44', '276', '1', '1000', '30', '18', '20'),
 ('96bd76ec8810374ed1b65e291975717f', 'esporte_lazer', '46', '250', '1', '154', '18', '9', '15'),
 ('cef67bcfe19066a932b7673e239eb23d', 'bebes', '27', '261', '1', '371', '26', '4', '26'),
 ('9dc1a7de274444849c219cff195d0b71', 'utilidades_domesticas', '37', '402', '4', '625', '20', '17', '13')]

In [ ]:
%%sql

ALTER TABLE order_items
ADD CONSTRAINT fk_name6
FOREIGN KEY (product_id)
REFERENCES products(product_id);

_______________________________________________________________________________________________________________________________

## <FONT color='green'>Order payments - Orders using order_id</FONT>


In [62]:
%%sql

SELECT * 
FROM order_payments
where order_id not in (SELECT order_id from orders);

 * mysql+pymysql://root:***@localhost:3306/brazil
0 rows affected.


[]

In [ ]:
%%sql

ALTER TABLE order_payments
ADD CONSTRAINT fk_name7
FOREIGN KEY (order_id)
REFERENCES orders(order_id);

_______________________________________________________________________________________________________________________________

## <FONT color='green'>Order reviews - Orders using order_id</FONT>


In [65]:
%%sql

SELECT *
FROM order_reviews
WHERE order_id NOT IN (SELECT order_id from orders)
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/brazil
0 rows affected.


[]

In [ ]:
%%sql


ALTER TABLE order_reviews
ADD CONSTRAINT fk_name8
FOREIGN KEY (order_id)
REFERENCES orders(order_id);

_______________________________________________________________________________________________________________________________

## <FONT color='green'>Products - Orders using seller_id</FONT>


In [67]:
%sql show tables;

 * mysql+pymysql://root:***@localhost:3306/brazil
9 rows affected.


[('category_translation',),
 ('customers',),
 ('geolocation',),
 ('order_items',),
 ('order_payments',),
 ('order_reviews',),
 ('orders',),
 ('products',),
 ('sellers',)]

In [68]:
%%sql 
SELECT * FROM category_translation limit 5;

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('agro_industria_e_comercio', 'agro_industry_and_commerce'),
 ('alimentos', 'food'),
 ('alimentos_bebidas', 'food_drink'),
 ('artes', 'art'),
 ('artes_e_artesanato', 'arts_and_craftmanship')]

In [69]:
%%sql 
SELECT * FROM products limit 5;

 * mysql+pymysql://root:***@localhost:3306/brazil
5 rows affected.


[('00066f42aeeb9f3007548bb9d3f33c38', 'perfumaria', '53', '596', '6', '300', '20', '16', '16'),
 ('00088930e925c41fd95ebfe695fd2655', 'automotivo', '56', '752', '4', '1225', '55', '10', '26'),
 ('0009406fd7479715e4bef61dd91f2462', 'cama_mesa_banho', '50', '266', '2', '300', '45', '15', '35'),
 ('000b8f95fcb9e0096488278317764d19', 'utilidades_domesticas', '25', '364', '3', '550', '19', '24', '12'),
 ('000d9be29b5207b54e86aa1b1ac54872', 'relogios_presentes', '48', '613', '4', '250', '22', '11', '15')]

In [70]:
%%sql 
SELECT DISTINCT(product_category_name)
FROM products
WHERE product_category_name not in (select product_category_name from category_translation)

 * mysql+pymysql://root:***@localhost:3306/brazil
2 rows affected.


[('pc_gamer',), ('portateis_cozinha_e_preparadores_de_alimentos',)]

In [ ]:
%%sql
-- Insert missing values in the translation table.
INSERT INTO category_translation 
VALUES
    ('pc_gamer', 'pc_gamer'),
    ('portateis_cozinha_e_preparadores_de_alimentos', 'portable_kitchen_food_preparators');

In [ ]:
%%sql

ALTER TABLE products
ADD CONSTRAINT fk_name9
FOREIGN KEY (product_category_name)
REFERENCES  category_translation(product_category_name);

# <font color='red'>Result Schema <font/>

<img src="./pic_of_new_schema.jpg" alt="Desired Schema"/>